src/main/java/org/embulk/input/marketo/delegate/ProgramInputPlugin.java

```diff
@@ -0,0 +1,219 @@
+        @ConfigDefault("null")
+        Optional<String> getTagType();
+
+        @Config("tag_value")
+        @ConfigDefault("null")
+        Optional<String> getTagVallue();
+
+        @Config("earliest_updated_at")
+        @ConfigDefault("null")
+        Optional<Date> getEarliestUpdatedAt();
+
+        @Config("latest_updated_at")
+        @ConfigDefault("null")
+        Optional<Date> getLatestUpdatedAt();
+
+        @Config("filter_type")
+        @ConfigDefault("null")
+        Optional<String> getFilterType();
+
+        @Config("filter_values")
+        @ConfigDefault("null")
+        Optional<List<String>> getFilterValues();
+
+        @Config("report_duration")
+        @ConfigDefault("null")
+        Optional<Long> getReportDuration();
+
+        void setLatestUpdatedAt(Optional<Date> latestUpdatedAt);
+    }
+
+    public ProgramInputPlugin()
+    {
+    }
+
+    @Override
+    public void validateInputTask(PluginTask task)
+    {
+        super.validateInputTask(task);
+        // validate if query_by is selected
+        if (task.getQueryBy().isPresent()) {
+            switch(task.getQueryBy().get()) {
+            case TAG_TYPE:
+                //make sure tag type and tag value are not empty
+                if (!task.getTagType().isPresent() || !task.getTagVallue().isPresent()) {
+                    throw new ConfigException("tag_type and tag_value are required when query by Tag Type");
+                }
+                break;
+            case DATE_RANGE:
+                // make sure earliest_updated_at is not empty
+                if (!task.getEarliestUpdatedAt().isPresent()) {
+                    throw new ConfigException("`earliest_updated_at` is required when query by Date Range");
+                }
+
+                if (task.getIncremental() && task.getLatestUpdatedAt().isPresent()) {
+                    DateTime latest = new DateTime(task.getLatestUpdatedAt().get());
+                    if (latest.isAfter(DateTime.now())) {
+                        throw new ConfigException(String.format("`latest_updated_at` (%s) cannot precede the current date (%s) when incremental import",
+                                        latest.toString(DATE_FORMATTER),
+                                        (DateTime.now().toString(DATE_FORMATTER))));
+                    }
+                }
+
+                DateTime earliest = new DateTime(task.getEarliestUpdatedAt().get());
+                if (task.getReportDuration().isPresent()) {
+                    logger.info("`report_duration` is present, Prefer `report_duration` over `latest_updated_at`");
+                    // Update the latestUpdatedAt for the config
+                    DateTime latest = earliest.plus(task.getReportDuration().get());
+                    // Only import until now for incremental import
+                    if (task.getIncremental() && latest.isAfter(DateTime.now())) {
+                        latest = DateTime.now();
```

Comment:
Should we "peacefully" show the message then stop execution when "now" is not the right time to run instead of re-assign latest? There is one downside of re-assign the latest is it will  change the beginning intention of duration of initial configuration.

For example, the initial configuration has the distance between earliest and latest is 10 days, and the scheduler runs once a day, it means the `report_duration` in the beginning will be 10 days but then will be changed to 1 day.

Reply:
yes, agree